lib/resty/etcd/cluster/healthcheck.lua

```diff
@@ -0,0 +1,205 @@
+        host = "",
+        type = "http",
+        req_headers = {"User-Agent: lua-resty-etcd"},
+        https_verify_certificate = true,
+        healthy = {
+            http_statuses = {200},
+            --active healthy health checks are disabled by default
+            interval = 0,
+            successes = 1,
+        },
+        unhealthy = {
+            http_statuses = {500, 501, 503, 502, 504, 505},
+            interval = 1,
+            http_failures = 1,
+            tcp_failures = 1,
+            timeouts = 1,
+        },
+    },
+    passive = {
+        type = "http",
+        healthy = {
+            http_statuses = {200, 201},
+            successes = 1,
+        },
+        unhealthy = {
+            http_statuses = {500},
+            http_failures = 1,
+            tcp_failures = 1,
+            --passive health checks are disabled by default
+            timeouts = 0,
+        },
+    },
+}
+
+
+local _M = {}
+
+
+function _M.report_failure(endpoint, osi, _, status)
+    if not checker then
+        return
+    end
+
+    if osi == "tcp" then
+        checker:report_tcp_failure(endpoint.host, tonumber(endpoint.port),
+                nil, nil, "active")
+        return
+    end
+
+    if osi == "http" then
+        if status >= 500 then
+            checker:report_http_status(endpoint.host, tonumber(endpoint.port),
+                    nil, status, "active")
+            return
+        end
+    end
+
+end
+
+
+function _M.fetch_health_nodes(endpoints)
+    if not checker then
+        return nil
+    end
+
+    for _, endpoint in ipairs(endpoints) do
+        local ok = checker:get_target_status(endpoint.host, tonumber(endpoint.port))
+        if ok then
+            return endpoint
+        end
+    end
+
+    utils.log_error("etcd cluster is unavailable")
+    return nil
+end
+
+
+function _M.checker_clear()
+    if checker then
+        checker:clear()
+    end
+end
+
+
+function _M.checker_stop()
+    if checker then
+        checker:stop()
+    end
+end
+
+
+function _M.run(opts, endpoints)
+    local shared_dict = ngx_shared[opts.cluster_healthcheck.shm_name]
+    if not shared_dict then
+        return nil, "failed to get ngx.shared dict: " .. opts.cluster_healthcheck.shm_name
+    end
+
+    --supported etcd version >= 3.3.0
+    --https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/monitoring.md#health-check
+    if not headthcheck_endpoint[opts.api_prefix] then
+        return nil,"unsupported health check for the etcd version < v3.3.0"
+    end
+
+    if not checker then
+        local ok, checks = pcall(tbl_copy_merge_defaults, opts.cluster_healthcheck.checks,
+                DEFAULTS)
+        if not ok then
+            utils.log_error("err: ", checks)
+            return nil, checks
+        end
+
+        if healthcheck == nil then
+            healthcheck = require("resty.healthcheck")
+        end
+
+        local err
+        checker, err = healthcheck.new({
```

Comment:
wow, I think that is the wrong way.

different etcd instance can create different checker object, the checker object should belong to the etcd instance.

so we can not use a shared `checker` for different etcd instance.